In [2]:
import numpy as np
import pandas as pd

In [3]:
df = pd.read_csv('tenet_movie_reviews_cleaned_dataset.csv')
df.head(3)

Unnamed: 0        author  \
0           0  paulclaassen   
1           1          neil   
2           2   pedroborges   

                                                 url        date  rating  \
0  https://www.imdb.com/review/rw8032959/?ref_=tt...  2022-04-04     5.0   
1  https://www.imdb.com/review/rw6050469/?ref_=tt...  2020-08-31     5.0   
2  https://www.imdb.com/review/rw6217957/?ref_=tt...  2020-10-30     7.0   

                                              review  year  review_length  \
0  Over-complicated, way too long, and not all th...  2022             58   
1                    Far too clever for its own good  2020             31   
2                                       Tenet Review  2020             12   

                                              scores  compound_scores  \
0  {'neg': 0.247, 'neu': 0.753, 'pos': 0.0, 'comp...          -0.3875   
1  {'neg': 0.0, 'neu': 0.459, 'pos': 0.541, 'comp...           0.7096   
2  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...           0.0000   

     labels  
0  negative  
1  positive  
2   neutral

In [4]:
df.columns

Index(['Unnamed: 0', 'author', 'url', 'date', 'rating', 'review', 'year',
       'review_length', 'scores', 'compound_scores', 'labels'],
      dtype='object')

In [5]:
df.drop('Unnamed: 0',axis=1,inplace=True)

In [6]:
df.head(2)

author                                                url  \
0  paulclaassen  https://www.imdb.com/review/rw8032959/?ref_=tt...   
1          neil  https://www.imdb.com/review/rw6050469/?ref_=tt...   

         date  rating                                             review  \
0  2022-04-04     5.0  Over-complicated, way too long, and not all th...   
1  2020-08-31     5.0                    Far too clever for its own good   

   year  review_length                                             scores  \
0  2022             58  {'neg': 0.247, 'neu': 0.753, 'pos': 0.0, 'comp...   
1  2020             31  {'neg': 0.0, 'neu': 0.459, 'pos': 0.541, 'comp...   

   compound_scores    labels  
0          -0.3875  negative  
1           0.7096  positive

In [7]:
#pip install spacy

In [1]:
import spacy

In [8]:
nlp = spacy.load('en_core_web_sm')

In [9]:
doc = nlp(df['review'][0])

In [17]:
doc.text

'Over-complicated, way too long, and not all that exciting.'

In [23]:
count = 0
print(doc.text)

for word in doc:
    if (word.pos_) == 'PUNCT':
        count = count+1
        print(word.pos_)
        print('--------------------')
print(count)

Over-complicated, way too long, and not all that exciting.
PUNCT
--------------------
PUNCT
--------------------
PUNCT
--------------------
PUNCT
--------------------
4


In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5178 entries, 0 to 5177
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   author           5178 non-null   object 
 1   url              5178 non-null   object 
 2   date             5178 non-null   object 
 3   rating           5178 non-null   float64
 4   review           5178 non-null   object 
 5   year             5178 non-null   int64  
 6   review_length    5178 non-null   int64  
 7   scores           5178 non-null   object 
 8   compound_scores  5178 non-null   float64
 9   labels           5178 non-null   object 
dtypes: float64(2), int64(2), object(6)
memory usage: 404.7+ KB


In [25]:
df['punct'] = 0
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5178 entries, 0 to 5177
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   author           5178 non-null   object 
 1   url              5178 non-null   object 
 2   date             5178 non-null   object 
 3   rating           5178 non-null   float64
 4   review           5178 non-null   object 
 5   year             5178 non-null   int64  
 6   review_length    5178 non-null   int64  
 7   scores           5178 non-null   object 
 8   compound_scores  5178 non-null   float64
 9   labels           5178 non-null   object 
 10  punct            5178 non-null   int64  
dtypes: float64(2), int64(3), object(6)
memory usage: 445.1+ KB


In [26]:
for i in range(len(df)):
    doc = nlp(df['review'][i])
    for word in doc:
        if (word.pos_) == 'PUNCT':
            df['punct'][i] = df['punct'][i] + 1

C:\Users\HP\AppData\Local\Temp\ipykernel_23796\1809218431.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['punct'][i] = df['punct'][i] + 1


In [27]:
df.head()

author                                                url  \
0  paulclaassen  https://www.imdb.com/review/rw8032959/?ref_=tt...   
1          neil  https://www.imdb.com/review/rw6050469/?ref_=tt...   
2   pedroborges  https://www.imdb.com/review/rw6217957/?ref_=tt...   
3      ThomDerd  https://www.imdb.com/review/rw6131705/?ref_=tt...   
4      dierregi  https://www.imdb.com/review/rw6051247/?ref_=tt...   

         date  rating                                             review  \
0  2022-04-04     5.0  Over-complicated, way too long, and not all th...   
1  2020-08-31     5.0                    Far too clever for its own good   
2  2020-10-30     7.0                                       Tenet Review   
3  2020-09-28     9.0         If you are into movies, Tenet is your jam.   
4  2020-08-31     1.0                           Pretentious action movie   

   year  review_length                                             scores  \
0  2022             58  {'neg': 0.247, 'neu': 0.753, 'pos': 0.0, 'comp...   
1  2020             31  {'neg': 0.0, 'neu': 0.459, 'pos': 0.541, 'comp...   
2  2020             12  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...   
3  2020             42  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...   
4  2020             24  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...   

   compound_scores    labels  punct  
0          -0.3875  negative      4  
1           0.7096  positive      0  
2           0.0000   neutral      0  
3           0.0000   neutral      2  
4           0.0000   neutral      0

In [28]:
x = df[['review_length','punct']]
y = df['labels']

In [29]:
import sklearn
from sklearn.model_selection import train_test_split

In [30]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=42)

### **LOGISTIC REGRESSION**

In [31]:
from sklearn.linear_model import LogisticRegression

In [32]:
lr_model = LogisticRegression(solver='lbfgs')

In [33]:
lr_model.fit(x_train,y_train)

LogisticRegression()

In [34]:
predictions = lr_model.predict(x_test)

In [35]:
from sklearn import metrics

In [39]:
data = pd.DataFrame(metrics.confusion_matrix(y_test,predictions),index=['neg','neu','pos'],columns=['neg','neu','pos'])
data

neg  neu  pos
neg    0  314  133
neu    0  476  133
pos    0  320  178

In [38]:
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

    negative       0.00      0.00      0.00       447
     neutral       0.43      0.78      0.55       609
    positive       0.40      0.36      0.38       498

    accuracy                           0.42      1554
   macro avg       0.28      0.38      0.31      1554
weighted avg       0.30      0.42      0.34      1554



C:\Users\HP\anaconda3\envs\dhruv-notes-datascience\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\anaconda3\envs\dhruv-notes-datascience\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\anaconda3\envs\dhruv-notes-datascience\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, m

In [40]:
metrics.accuracy_score(y_test,predictions)

0.42084942084942084

### **NAIVE BAYES**

In [45]:
from sklearn.naive_bayes import MultinomialNB

In [46]:
nb = MultinomialNB()
nb.fit(x_train,y_train)
predictions = nb.predict(x_test)

In [47]:
data = pd.DataFrame(metrics.confusion_matrix(y_test,predictions),index=['neg','neu','pos'],columns=['neg','neu','pos'])
data

neg  neu  pos
neg    7  432    8
neu   10  593    6
pos   14  463   21

In [48]:
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

    negative       0.23      0.02      0.03       447
     neutral       0.40      0.97      0.57       609
    positive       0.60      0.04      0.08       498

    accuracy                           0.40      1554
   macro avg       0.41      0.34      0.22      1554
weighted avg       0.41      0.40      0.26      1554



In [49]:
metrics.accuracy_score(y_test,predictions)

0.3996138996138996

### **SUPPORT VECTOR MECHANISM**

In [50]:
from sklearn.svm import SVC

In [51]:
svc_model = SVC()
svc_model.fit(x_train,y_train)
predictions = svc_model.predict(x_test)

In [52]:
data = pd.DataFrame(metrics.confusion_matrix(y_test,predictions),index=['neg','neu','pos'],columns=['neg','neu','pos'])
data

neg  neu  pos
neg    0  333  114
neu    0  498  111
pos    0  337  161

In [53]:
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

    negative       0.00      0.00      0.00       447
     neutral       0.43      0.82      0.56       609
    positive       0.42      0.32      0.36       498

    accuracy                           0.42      1554
   macro avg       0.28      0.38      0.31      1554
weighted avg       0.30      0.42      0.34      1554



C:\Users\HP\anaconda3\envs\dhruv-notes-datascience\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\anaconda3\envs\dhruv-notes-datascience\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\anaconda3\envs\dhruv-notes-datascience\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, m

In [54]:
metrics.accuracy_score(y_test,predictions)

0.4240669240669241

## **TEXT ANALYTICS**
### **TF-IDF Vectorizer and LINEAR Support Vector Classifier**

In [55]:
x = df['review']
y = df['labels']

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=42)

In [56]:
x_train.shape,y_train.shape

((3624,), (3624,))

In [57]:
x_test.shape,y_test.shape

((1554,), (1554,))

In [58]:
from sklearn.pipeline import Pipeline

In [59]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [60]:
from sklearn.svm import LinearSVC

In [65]:
text_clf = Pipeline([('tfidf',TfidfVectorizer()),('clf',LinearSVC())])

In [66]:
text_clf.fit(x_train,y_train)

Pipeline(steps=[('tfidf', TfidfVectorizer()), ('clf', LinearSVC())])

In [67]:
predictions = text_clf.predict(x_test)

In [68]:
x_test[0:5]

1569                                  Inconsistently dumb
751               Terrible audio, vomit inducing visuals.
3017          Nolan is changing the way we "view" movies.
4223    A hollywood checklist of criteria combined wit...
4001    Thank you, Mr Nolan, you are not making it eas...
Name: review, dtype: object

In [69]:
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

    negative       0.82      0.74      0.78       447
     neutral       0.81      0.89      0.84       609
    positive       0.82      0.80      0.81       498

    accuracy                           0.82      1554
   macro avg       0.82      0.81      0.81      1554
weighted avg       0.82      0.82      0.81      1554



In [70]:
data = pd.DataFrame(metrics.confusion_matrix(y_test,predictions),index=['neg','neu','pos'],columns=['neg','neu','pos'])
data

neg  neu  pos
neg  331   65   51
neu   35  539   35
pos   37   64  397

In [71]:
metrics.accuracy_score(y_test,predictions)

0.8153153153153153

In [73]:
text_clf.predict(['Over-complicated, way too long, and not all that exciting.'])

array(['negative'], dtype=object)

-------------
-------------

### **Accuracy**

**Logistic Regression** = 0.424

**Naive Bayes** = 0.39

**Support Vector Classifier** = 0.424

**Text Analytics = TF-IDF and Linear SVC** = 0.81